File: service/app/src/test/java/io/miragon/blueprint/process/BikeLeasingProcessTest.java
package io.miragon.blueprint.process;

import static io.miragon.blueprint.process.util.JobExecutionUtils.continueToNextWaitState;
import static io.miragon.blueprint.process.util.JobExecutionUtils.executeJobFor;
import static io.miragon.blueprint.process.util.ProcessInstanceUtils.findProcessInstance;
import static io.miragon.blueprint.process.util.TimerUtils.fireTimer;
import static org.operaton.bpm.engine.test.assertions.bpmn.BpmnAwareTests.assertThat;
import static org.operaton.bpm.engine.test.assertions.bpmn.BpmnAwareTests.init;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.miragon.blueprint.adapter.process.BikeLeasingProcessProcessApi.Elements;
import io.miragon.blueprint.adapter.process.CancelBikeOrderProcessApi;
import io.miragon.blueprint.application.port.inbound.ActivateLeasingUseCase;
import io.miragon.blueprint.application.port.inbound.BookCancellationCostsUseCase;
import io.miragon.blueprint.application.port.inbound.CancelContractUseCase;
import io.miragon.blueprint.application.port.inbound.CancelInsurancePolicyUseCase;
import io.miragon.blueprint.application.port.inbound.IssueInsurancePolicyUseCase;
import io.miragon.blueprint.application.port.inbound.OrderBikeUseCase;
import io.miragon.blueprint.application.port.inbound.RejectApplicationUseCase;
import io.miragon.blueprint.application.port.inbound.RequestOrderCancellationUseCase;
import io.miragon.blueprint.application.port.inbound.SendCancellationConfirmationUseCase;
import io.miragon.blueprint.application.port.inbound.SendContractUseCase;
import io.miragon.blueprint.application.port.inbound.SendSignatureReminderUseCase;
import io.miragon.blueprint.application.port.inbound.ValidateApplicationUseCase;
import io.miragon.blueprint.application.port.outbound.LeasingProcess;
import io.miragon.blueprint.domain.bike.BikeId;
import io.miragon.blueprint.domain.bike.OrderId;
import io.miragon.blueprint.domain.leasing.ApplicationId;
import io.miragon.blueprint.domain.leasing.CustomerName;
import io.miragon.blueprint.domain.leasing.Email;
import io.miragon.blueprint.domain.leasing.LeasingApplication;
import io.miragon.blueprint.domain.leasing.LeasingStatus;
import java.time.LocalDateTime;
import java.util.Map;
import org.operaton.bpm.engine.ProcessEngine;
import org.operaton.bpm.engine.RuntimeService;
import org.operaton.bpm.engine.TaskService;
import org.operaton.bpm.engine.runtime.ProcessInstance;
import org.operaton.bpm.engine.task.Task;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.bean.override.mockito.MockitoBean;

@SpringBootTest
@ActiveProfiles("test")
class BikeLeasingProcessTest {

    @Autowired
    private LeasingProcess process;

    @Autowired
    private RuntimeService runtimeService;

    @Autowired
    private TaskService taskService;

    @Autowired
    private ProcessEngine processEngine;

    @MockitoBean
    private ValidateApplicationUseCase validateApplicationUseCase;

    @MockitoBean
    private RejectApplicationUseCase rejectApplicationUseCase;

    @MockitoBean
    private SendContractUseCase sendContractUseCase;

    @MockitoBean
    private CancelContractUseCase cancelContractUseCase;

    @MockitoBean
    private IssueInsurancePolicyUseCase issueInsurancePolicyUseCase;

    @MockitoBean
    private CancelInsurancePolicyUseCase cancelInsurancePolicyUseCase;

    @MockitoBean
    private SendSignatureReminderUseCase sendSignatureReminderUseCase;

    @MockitoBean
    private SendCancellationConfirmationUseCase sendCancellationConfirmationUseCase;

    @MockitoBean
    private RequestOrderCancellationUseCase requestOrderCancellationUseCase;

    @MockitoBean
    private BookCancellationCostsUseCase bookCancellationCostsUseCase;

    @MockitoBean
    private OrderBikeUseCase orderBikeUseCase;

    @MockitoBean
    private ActivateLeasingUseCase activateLeasingUseCase;

    @BeforeEach
    void setUp() {
        init(processEngine);
        when(orderBikeUseCase.orderBike(any()))
                .thenReturn(new OrderBikeUseCase.Result(new OrderId("ORDER-1"), true));
    }

    @Test
    void happyPathContractSignedBikeAvailableLeasingBecomesActive() {
        ApplicationId id = submit(35, 3500.0);
        ProcessInstance instance = findProcessInstance(runtimeService, id);

        // async continuations up to the contract-signature wait state
        executeJobFor(processEngine, Elements.START_EVENT_LEASING_REQUEST_RECEIVED);
        executeJobFor(processEngine, Elements.SERVICE_TASK_VALIDATE_APPLICATION);
        executeJobFor(processEngine, Elements.SERVICE_TASK_SEND_CONTRACT);

        process.correlateContractSigned(id);
        executeJobFor(processEngine, Elements.EVENT_CONTRACT_SIGNED); // forks into insurance + bike order
        executeJobFor(processEngine, Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY);
        executeJobFor(processEngine, Elements.SERVICE_TASK_ORDER_BIKE); // joins -> handover wait state

        process.correlateHandoverReported(id);
        executeJobFor(processEngine, Elements.EVENT_HANDOVER_REPORTED); // -> withdrawal-period timer

        fireTimer(processEngine, Elements.EVENT_WITHDRAWAL_PERIOD_ELAPSED);

        assertThat(instance)
                .isEnded()
                .hasPassedInOrder(
                        Elements.SERVICE_TASK_VALIDATE_APPLICATION.getValue(),
                        Elements.BUSINESS_RULE_TASK_CHECK_CREDIT_RATING.getValue(),
                        Elements.SERVICE_TASK_SEND_CONTRACT.getValue(),
                        Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY.getValue(),
                        Elements.EVENT_HANDOVER_REPORTED.getValue(),
                        Elements.END_EVENT_LEASING_ACTIVE.getValue())
                .hasNotPassed(
                        Elements.END_EVENT_APPLICATION_REJECTED.getValue(),
                        Elements.END_EVENT_APPLICATION_CANCELLED.getValue(),
                        Elements.END_EVENT_CONTRACT_CANCELLED.getValue());

        verify(sendContractUseCase, times(1)).sendContract(id);
        verify(issueInsurancePolicyUseCase, times(1)).issuePolicy(id);
        verify(activateLeasingUseCase, times(1)).activate(id);
    }

    @Test
    void escalationContractNotSignedInTimeIsEscalatedAndRejected() {
        ApplicationId id = submit(35, 3500.0);
        ProcessInstance instance = findProcessInstance(runtimeService, id);

        // async continuations up to the contract-signature wait state
        executeJobFor(processEngine, Elements.START_EVENT_LEASING_REQUEST_RECEIVED);
        executeJobFor(processEngine, Elements.SERVICE_TASK_VALIDATE_APPLICATION);
        executeJobFor(processEngine, Elements.SERVICE_TASK_SEND_CONTRACT);

        fireTimer(processEngine, Elements.EVENT_SIGNATURE_DEADLINE); // deadline -> escalation -> rejection
        executeJobFor(processEngine, Elements.SERVICE_TASK_SEND_REJECTION);

        assertThat(instance)
                .isEnded()
                .hasPassed(
                        Elements.EVENT_SIGNATURE_DEADLINE.getValue(),
                        Elements.EVENT_CONTRACT_NOT_SIGNED.getValue(),
                        Elements.SERVICE_TASK_SEND_REJECTION.getValue(),
                        Elements.END_EVENT_APPLICATION_REJECTED.getValue())
                .hasNotPassed(Elements.END_EVENT_LEASING_ACTIVE.getValue());

        verify(rejectApplicationUseCase, times(1)).reject(id);
    }

    @Test
    void notSolventTheDmnRoutesTheApplicationStraightToRejection() {
        // age below 18 cannot sign a leasing contract, so the DMN returns solvent = false
        ApplicationId id = submit(15, 3500.0);

        executeJobFor(processEngine, Elements.START_EVENT_LEASING_REQUEST_RECEIVED);
        executeJobFor(processEngine, Elements.SERVICE_TASK_VALIDATE_APPLICATION); // -> DMN -> not solvent -> rejection
        executeJobFor(processEngine, Elements.SERVICE_TASK_SEND_REJECTION);

        verify(rejectApplicationUseCase, times(1)).reject(id);
        verify(sendContractUseCase, never()).sendContract(any());
    }

    @Test
    void abortWithdrawingTheApplicationCompensatesTheCompletedSteps() {
        when(requestOrderCancellationUseCase.requestCancellation(any())).thenReturn(true);

        ApplicationId id = submit(35, 3500.0);
        ProcessInstance instance = findProcessInstance(runtimeService, id);

        // async continuations up to the handover wait state (contract signed, bike ordered, insured)
        executeJobFor(processEngine, Elements.START_EVENT_LEASING_REQUEST_RECEIVED);
        executeJobFor(processEngine, Elements.SERVICE_TASK_VALIDATE_APPLICATION);
        executeJobFor(processEngine, Elements.SERVICE_TASK_SEND_CONTRACT);
        process.correlateContractSigned(id);
        executeJobFor(processEngine, Elements.EVENT_CONTRACT_SIGNED);
        executeJobFor(processEngine, Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY);
        executeJobFor(processEngine, Elements.SERVICE_TASK_ORDER_BIKE);

        // Withdrawing triggers compensation. Its handlers run in an engine-defined order, so drive
        // the continuations generically until the cancelBikeOrder sub-process parks on its user task.
        process.correlateApplicationWithdrawn(id);
        continueToNextWaitState(processEngine);

        Task task = taskService
                .createTaskQuery()
                .taskDefinitionKey(CancelBikeOrderProcessApi.Elements.USER_TASK_CLARIFY_RETURN.getValue())
                .singleResult();
        taskService.complete(task.getId(), Map.of("returnClarified", true));
        continueToNextWaitState(processEngine);

        assertThat(instance)
                .isEnded()
                .hasPassed(
                        Elements.SERVICE_TASK_CANCEL_CONTRACT.getValue(),
                        Elements.SERVICE_TASK_CANCEL_POLICY.getValue(),
                        Elements.CALL_ACTIVITY_CANCEL_BIKE_ORDER.getValue(),
                        Elements.SERVICE_TASK_SEND_CANCELLATION_CONFIRMATION.getValue(),
                        Elements.END_EVENT_APPLICATION_CANCELLED.getValue())
                .hasNotPassed(Elements.END_EVENT_LEASING_ACTIVE.getValue());

        verify(cancelContractUseCase, times(1)).cancelContract(id);
        verify(cancelInsurancePolicyUseCase, times(1)).cancelPolicy(id);
        verify(sendCancellationConfirmationUseCase, times(1)).sendCancellationConfirmation(id);
    }

    @Test
    void bikeUnavailableClarifyingAnAlternativeReOrdersAndLeasingBecomesActive() {
        // the first order finds the requested bike unavailable, the re-order after the alternative succeeds
        when(orderBikeUseCase.orderBike(any())).thenReturn(
                new OrderBikeUseCase.Result(null, false),
                new OrderBikeUseCase.Result(new OrderId("ORDER-2"), true));

        ApplicationId id = submit(35, 3500.0);
        ProcessInstance instance = findProcessInstance(runtimeService, id);

        // async continuations up to the contract-signature wait state, then fork into insurance + bike order
        executeJobFor(processEngine, Elements.START_EVENT_LEASING_REQUEST_RECEIVED);
        executeJobFor(processEngine, Elements.SERVICE_TASK_VALIDATE_APPLICATION);
        executeJobFor(processEngine, Elements.SERVICE_TASK_SEND_CONTRACT);
        process.correlateContractSigned(id);
        executeJobFor(processEngine, Elements.EVENT_CONTRACT_SIGNED);
        executeJobFor(processEngine, Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY);
        executeJobFor(processEngine, Elements.SERVICE_TASK_ORDER_BIKE); // unavailable -> parks on clarify-alternative

        // the alternative is clarified from the outside — the "external" completion of the user task
        process.completeAlternativeClarification(id, true, new BikeId("BIKE-ALT"));
        continueToNextWaitState(processEngine); // re-order succeeds -> parallel join -> handover wait state

        process.correlateHandoverReported(id);
        executeJobFor(processEngine, Elements.EVENT_HANDOVER_REPORTED);
        fireTimer(processEngine, Elements.EVENT_WITHDRAWAL_PERIOD_ELAPSED);

        assertThat(instance)
                .isEnded()
                .hasPassed(
                        Elements.USER_TASK_CLARIFY_ALTERNATIVE.getValue(),
                        Elements.SERVICE_TASK_ORDER_BIKE.getValue(),
                        Elements.END_EVENT_LEASING_ACTIVE.getValue())
                .hasNotPassed(
                        Elements.END_EVENT_CONTRACT_CANCELLED.getValue(),
                        Elements.END_EVENT_APPLICATION_REJECTED.getValue());

        verify(orderBikeUseCase, times(2)).orderBike(id);
        verify(activateLeasingUseCase, times(1)).activate(id);
    }

    private ApplicationId submit(int age, double income) {
        LeasingApplication application = new LeasingApplication(
                ApplicationId.newId(),
                new CustomerName("Test Customer"),
                new Email("test@example.com"),
                age,
                income,
                new BikeId("BIKE-TEST"),
                LeasingStatus.RECEIVED,
                LocalDateTime.now(),
                null,
                null);
        process.submitRequest(application);
        return application.id();
    }
}

File: service/app/src/test/kotlin/io/miragon/blueprint/process/BikeLeasingProcessTest.kt
package io.miragon.blueprint.process

import com.ninjasquad.springmockk.MockkBean
import io.miragon.blueprint.adapter.process.BikeLeasingProcessProcessApi.Elements
import io.miragon.blueprint.adapter.process.CancelBikeOrderProcessApi
import io.miragon.blueprint.application.port.inbound.BookCancellationCostsUseCase
import io.miragon.blueprint.application.port.inbound.CancelContractUseCase
import io.miragon.blueprint.application.port.inbound.CancelInsurancePolicyUseCase
import io.miragon.blueprint.application.port.inbound.IssueInsurancePolicyUseCase
import io.miragon.blueprint.application.port.inbound.OrderBikeUseCase
import io.miragon.blueprint.application.port.inbound.RejectApplicationUseCase
import io.miragon.blueprint.application.port.inbound.RequestOrderCancellationUseCase
import io.miragon.blueprint.application.port.inbound.SendCancellationConfirmationUseCase
import io.miragon.blueprint.application.port.inbound.SendContractUseCase
import io.miragon.blueprint.application.port.inbound.SendSignatureReminderUseCase
import io.miragon.blueprint.application.port.inbound.ValidateApplicationUseCase
import io.miragon.blueprint.application.port.outbound.LeasingProcess
import io.miragon.blueprint.domain.leasing.ApplicationId
import io.miragon.blueprint.domain.bike.BikeId
import io.miragon.blueprint.domain.leasing.CustomerName
import io.miragon.blueprint.domain.leasing.Email
import io.miragon.blueprint.domain.leasing.LeasingApplication
import io.miragon.blueprint.domain.leasing.LeasingStatus
import io.miragon.blueprint.domain.bike.OrderId
import io.miragon.blueprint.process.util.continueToNextWaitState
import io.miragon.blueprint.process.util.executeJobFor
import io.miragon.blueprint.process.util.findProcessInstance
import io.miragon.blueprint.process.util.fireTimer
import io.mockk.every
import io.mockk.verify
import org.operaton.bpm.engine.ProcessEngine
import org.operaton.bpm.engine.RuntimeService
import org.operaton.bpm.engine.TaskService
import org.operaton.bpm.engine.test.assertions.bpmn.BpmnAwareTests.assertThat
import org.operaton.bpm.engine.test.assertions.bpmn.BpmnAwareTests.init
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import java.time.LocalDateTime

@SpringBootTest
@ActiveProfiles("test")
class BikeLeasingProcessTest {

    @Autowired
    private lateinit var process: LeasingProcess

    @Autowired
    private lateinit var runtimeService: RuntimeService

    @Autowired
    private lateinit var taskService: TaskService

    @Autowired
    private lateinit var processEngine: ProcessEngine

    @MockkBean(relaxed = true)
    private lateinit var validateApplicationUseCase: ValidateApplicationUseCase

    @MockkBean(relaxed = true)
    private lateinit var rejectApplicationUseCase: RejectApplicationUseCase

    @MockkBean(relaxed = true)
    private lateinit var sendContractUseCase: SendContractUseCase

    @MockkBean(relaxed = true)
    private lateinit var cancelContractUseCase: CancelContractUseCase

    @MockkBean(relaxed = true)
    private lateinit var issueInsurancePolicyUseCase: IssueInsurancePolicyUseCase

    @MockkBean(relaxed = true)
    private lateinit var cancelInsurancePolicyUseCase: CancelInsurancePolicyUseCase

    @MockkBean(relaxed = true)
    private lateinit var sendSignatureReminderUseCase: SendSignatureReminderUseCase

    @MockkBean(relaxed = true)
    private lateinit var sendCancellationConfirmationUseCase: SendCancellationConfirmationUseCase

    @MockkBean(relaxed = true)
    private lateinit var requestOrderCancellationUseCase: RequestOrderCancellationUseCase

    @MockkBean(relaxed = true)
    private lateinit var bookCancellationCostsUseCase: BookCancellationCostsUseCase

    @MockkBean(relaxed = true)
    private lateinit var orderBikeUseCase: OrderBikeUseCase

    @BeforeEach
    fun setUp() {
        init(processEngine)
        every { orderBikeUseCase.orderBike(any()) } returns
            OrderBikeUseCase.Result(OrderId("ORDER-1"), bikeAvailable = true)
    }

    @Test
    fun `happy path - contract signed, bike available, leasing becomes active`() {
        val id = submit(age = 35, income = 3500.0)
        val instance = runtimeService.findProcessInstance(id)

        // async continuations up to the contract-signature wait state
        processEngine.executeJobFor(Elements.START_EVENT_LEASING_REQUEST_RECEIVED)
        processEngine.executeJobFor(Elements.SERVICE_TASK_VALIDATE_APPLICATION)
        processEngine.executeJobFor(Elements.SERVICE_TASK_SEND_CONTRACT)

        process.correlateContractSigned(id)
        processEngine.executeJobFor(Elements.EVENT_CONTRACT_SIGNED) // forks into insurance + bike order
        processEngine.executeJobFor(Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY)
        processEngine.executeJobFor(Elements.SERVICE_TASK_ORDER_BIKE) // joins -> handover wait state

        process.correlateHandoverReported(id)
        processEngine.executeJobFor(Elements.EVENT_HANDOVER_REPORTED) // -> withdrawal-period timer

        processEngine.fireTimer(Elements.EVENT_WITHDRAWAL_PERIOD_ELAPSED)

        assertThat(instance)
            .isEnded
            .hasPassedInOrder(
                Elements.SERVICE_TASK_VALIDATE_APPLICATION.value,
                Elements.BUSINESS_RULE_TASK_CHECK_CREDIT_RATING.value,
                Elements.SERVICE_TASK_SEND_CONTRACT.value,
                Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY.value,
                Elements.EVENT_HANDOVER_REPORTED.value,
                Elements.END_EVENT_LEASING_ACTIVE.value,
            )
            .hasNotPassed(
                Elements.END_EVENT_APPLICATION_REJECTED.value,
                Elements.END_EVENT_APPLICATION_CANCELLED.value,
                Elements.END_EVENT_CONTRACT_CANCELLED.value,
            )

        verify(exactly = 1) { sendContractUseCase.sendContract(id) }
        verify(exactly = 1) { issueInsurancePolicyUseCase.issuePolicy(id) }
    }

    @Test
    fun `escalation - contract not signed in time is escalated and rejected`() {
        val id = submit(age = 35, income = 3500.0)
        val instance = runtimeService.findProcessInstance(id)

        // async continuations up to the contract-signature wait state
        processEngine.executeJobFor(Elements.START_EVENT_LEASING_REQUEST_RECEIVED)
        processEngine.executeJobFor(Elements.SERVICE_TASK_VALIDATE_APPLICATION)
        processEngine.executeJobFor(Elements.SERVICE_TASK_SEND_CONTRACT)

        processEngine.fireTimer(Elements.EVENT_SIGNATURE_DEADLINE) // deadline -> escalation -> rejection
        processEngine.executeJobFor(Elements.SERVICE_TASK_SEND_REJECTION)

        assertThat(instance)
            .isEnded
            .hasPassed(
                Elements.EVENT_SIGNATURE_DEADLINE.value,
                Elements.BOUNDARY_CONTRACT_NOT_SIGNED.value,
                Elements.SERVICE_TASK_SEND_REJECTION.value,
                Elements.END_EVENT_APPLICATION_REJECTED.value,
            )
            .hasNotPassed(Elements.END_EVENT_LEASING_ACTIVE.value)

        verify(exactly = 1) { rejectApplicationUseCase.reject(id) }
    }

    @Test
    fun `not solvent - the DMN routes the application straight to rejection`() {
        // age below 18 cannot sign a leasing contract, so the DMN returns solvent = false
        val id = submit(age = 15, income = 3500.0)

        processEngine.executeJobFor(Elements.START_EVENT_LEASING_REQUEST_RECEIVED)
        processEngine.executeJobFor(Elements.SERVICE_TASK_VALIDATE_APPLICATION) // -> DMN -> not solvent -> rejection
        processEngine.executeJobFor(Elements.SERVICE_TASK_SEND_REJECTION)

        verify(exactly = 1) { rejectApplicationUseCase.reject(id) }
        verify(exactly = 0) { sendContractUseCase.sendContract(any()) }
    }

    @Test
    fun `abort - withdrawing the application compensates the completed steps`() {
        every { requestOrderCancellationUseCase.requestCancellation(any()) } returns true

        val id = submit(age = 35, income = 3500.0)
        val instance = runtimeService.findProcessInstance(id)

        // async continuations up to the handover wait state (contract signed, bike ordered, insured)
        processEngine.executeJobFor(Elements.START_EVENT_LEASING_REQUEST_RECEIVED)
        processEngine.executeJobFor(Elements.SERVICE_TASK_VALIDATE_APPLICATION)
        processEngine.executeJobFor(Elements.SERVICE_TASK_SEND_CONTRACT)
        process.correlateContractSigned(id)
        processEngine.executeJobFor(Elements.EVENT_CONTRACT_SIGNED)
        processEngine.executeJobFor(Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY)
        processEngine.executeJobFor(Elements.SERVICE_TASK_ORDER_BIKE)

        // Withdrawing triggers compensation. Its handlers run in an engine-defined order, so drive
        // the continuations generically until the cancelBikeOrder sub-process parks on its user task.
        process.correlateApplicationWithdrawn(id)
        processEngine.continueToNextWaitState()

        val task =
            taskService
                .createTaskQuery()
                .taskDefinitionKey(CancelBikeOrderProcessApi.Elements.USER_TASK_CLARIFY_RETURN.value)
                .singleResult()
        taskService.complete(task.id, mapOf("returnClarified" to true))
        processEngine.continueToNextWaitState()

        assertThat(instance)
            .isEnded
            .hasPassed(
                Elements.SERVICE_TASK_CANCEL_CONTRACT.value,
                Elements.SERVICE_TASK_CANCEL_POLICY.value,
                Elements.CALL_ACTIVITY_CANCEL_BIKE_ORDER.value,
                Elements.SERVICE_TASK_SEND_CANCELLATION_CONFIRMATION.value,
                Elements.END_EVENT_APPLICATION_CANCELLED.value,
            )
            .hasNotPassed(Elements.END_EVENT_LEASING_ACTIVE.value)

        verify(exactly = 1) { cancelContractUseCase.cancelContract(id) }
        verify(exactly = 1) { cancelInsurancePolicyUseCase.cancelPolicy(id) }
        verify(exactly = 1) { sendCancellationConfirmationUseCase.sendCancellationConfirmation(id) }
    }

    @Test
    fun `bike unavailable - clarifying an alternative re-orders and leasing becomes active`() {
        // the first order finds the requested bike unavailable, the re-order after the alternative succeeds
        every { orderBikeUseCase.orderBike(any()) } returnsMany
            listOf(
                OrderBikeUseCase.Result(orderId = null, bikeAvailable = false),
                OrderBikeUseCase.Result(OrderId("ORDER-2"), bikeAvailable = true),
            )

        val id = submit(age = 35, income = 3500.0)
        val instance = runtimeService.findProcessInstance(id)

        // async continuations up to the contract-signature wait state, then fork into insurance + bike order
        processEngine.executeJobFor(Elements.START_EVENT_LEASING_REQUEST_RECEIVED)
        processEngine.executeJobFor(Elements.SERVICE_TASK_VALIDATE_APPLICATION)
        processEngine.executeJobFor(Elements.SERVICE_TASK_SEND_CONTRACT)
        process.correlateContractSigned(id)
        processEngine.executeJobFor(Elements.EVENT_CONTRACT_SIGNED)
        processEngine.executeJobFor(Elements.SERVICE_TASK_ISSUE_INSURANCE_POLICY)
        processEngine.executeJobFor(Elements.SERVICE_TASK_ORDER_BIKE) // unavailable -> parks on the clarify-alternative user task

        // the alternative is clarified from the outside — the "external" completion of the user task
        process.completeAlternativeClarification(id, alternativeFound = true, bikeId = BikeId("BIKE-ALT"))
        processEngine.continueToNextWaitState() // re-order succeeds -> parallel join -> handover wait state

        process.correlateHandoverReported(id)
        processEngine.executeJobFor(Elements.EVENT_HANDOVER_REPORTED)
        processEngine.fireTimer(Elements.EVENT_WITHDRAWAL_PERIOD_ELAPSED)

        assertThat(instance)
            .isEnded
            .hasPassed(
                Elements.USER_TASK_CLARIFY_ALTERNATIVE.value,
                Elements.SERVICE_TASK_ORDER_BIKE.value,
                Elements.END_EVENT_LEASING_ACTIVE.value,
            )
            .hasNotPassed(
                Elements.END_EVENT_CONTRACT_CANCELLED.value,
                Elements.END_EVENT_APPLICATION_REJECTED.value,
            )

        verify(exactly = 2) { orderBikeUseCase.orderBike(id) }
    }

    private fun submit(age: Int, income: Double, bikeId: String = "BIKE-TEST"): ApplicationId {
        val application =
            LeasingApplication(
                id = ApplicationId.new(),
                customerName = CustomerName("Test Customer"),
                email = Email("test@example.com"),
                age = age,
                monthlyNetIncome = income,
                bikeId = BikeId(bikeId),
                status = LeasingStatus.RECEIVED,
                createdAt = LocalDateTime.now(),
            )
        process.submitRequest(application)
        return application.id
    }
}
